engine/src/widget-events.cpp

```diff
@@ -496,26 +508,121 @@ void MCWidgetEventManager::event_paint(MCWidget *p_widget, MCGContextRef p_gcont
     MCWidgetOnPaint(p_widget -> getwidget(), p_gcontext);
 }
 
-void MCWidgetEventManager::event_touch(MCWidget* p_widget, uint32_t p_id, MCEventTouchPhase p_phase, int2 p_x, int2 p_y)
+Bool MCWidgetEventManager::event_touch(MCWidget* p_widget, uint32_t p_id, MCEventTouchPhase p_phase, int2 p_x, int2 p_y)
 {
+    if (!widgetIsInRunMode(p_widget))
+    {
+        return False;
+    }
+    
+    /* If a touch is just starting, and this is the first touch then we need
+     * to compute the touched widget. */
+    MCWidgetRef t_touched_widget = m_touched_widget;
+    if (p_phase == kMCEventTouchPhaseBegan)
+    {
+        if (t_touched_widget == nullptr)
+        {
+            t_touched_widget = hitTest(p_widget->getwidget(), p_x, p_y);
+        }
+    }
+    
+    /* If we don't have a touched widget, then ignore the touch. */
+    if (t_touched_widget == nullptr)
+    {
+        return False;
+    }
+    
+    /* See if there is already a slot for the given touch. If none is found
+     * and this is not a begin, we ignore it - otherwise we create one. */
+    uinteger_t t_slot;
+    if (!findTouchSlotById(p_id, t_slot))
+    {
+        if (p_phase == kMCEventTouchPhaseBegan)
+        {
+            t_slot = allocateTouchSlot();
+            m_touches[t_slot].m_id = p_id;
+        }
+        else
+        {
+            return False;
+        }
+    }
+    else if (p_phase == kMCEventTouchPhaseBegan)
+    {
+        /* This shouldn't happen - it would mean we are getting two began
```

Comment:
@runrevmark shouldn't we `MCUnreachable` this?

Reply:
I put that in because it was in the original code - we could put an MCAssert or such although this is a fault-tolerant resolution so I'm not sure it matters too much.